tsdb/engine/tsm1/cache.go

```diff
@@ -180,37 +187,79 @@ func (c *Cache) Keys() []string {
 
 // Values returns a copy of all values, deduped and sorted, for the given key.
 func (c *Cache) Values(key string) Values {
```

Comment:
I wonder if a copy-on-write approach might be better in the long run. In other words when a query comes in, and assuming the data is sorted, the query gets a reference to the hot data, *and* we bump a reference counter. Now, if more data comes in and the reference count is non-zero, only then do we make the copy, but the copy is made for the write traffic, and that copy then becomes the hot data (the queries continue to use their own, now stale, copy).

If we avoid copying both snapshotted-data and hot data, it would mean queries should never need copies of the cache (except in the one case that a sort is needed). Two queries back-to-back, without any writes taking place to that series, would both be zero-copy.

This would mean a write-lock would always need to be taken (to bump the reference counter) but it could be a win overall.

Not critical right now, but something we can consider.